File: redo/osxaltest/box.m
// 31 july 2015
#import "osxaltest.h"

// leave a whole lot of space around the alignment rect, just to be safe
// TODO fine tune this
// TODO de-duplicate this from spinbox.m
@interface tBoxContainer : NSView
@end

@implementation tBoxContainer

- (NSEdgeInsets)alignmentRectInsets
{
	return NSEdgeInsetsMake(50, 50, 50, 50);
}

@end

@implementation tBox {
	NSView *v;
	NSMutableArray *children;
	NSMutableArray *stretchy;
	BOOL vertical;
	id<tControl> parent;
	BOOL spaced;
}

// TODO rename to padded
- (id)tInitVertical:(BOOL)vert spaced:(BOOL)sp
{
	self = [super init];
	if (self) {
		self->v = [[NSView alloc] initWithFrame:NSZeroRect];
		[self->v setTranslatesAutoresizingMaskIntoConstraints:NO];
		self->children = [NSMutableArray new];
		self->stretchy = [NSMutableArray new];
		self->vertical = vert;
		self->parent = nil;
		self->spaced = sp;
	}
	return self;
}

- (void)tAddControl:(id<tControl>)c stretchy:(BOOL)s
{
	[c tSetParent:self addToView:self->v relayout:NO];
	[self->children addObject:c];
	[self->stretchy addObject:[NSNumber numberWithBool:s]];
	[self tRelayout];
}

- (void)tSetParent:(id<tControl>)p addToView:(NSView *)sv relayout:(BOOL)relayout
{
	self->parent = p;
	[sv addSubview:self->v];
	if (relayout)
		[self tRelayout];
}

// TODO custom minimum width for non-stretchy controls
- (void)tFillAutoLayout:(tAutoLayoutParams *)p
{
	NSMutableDictionary *views;
	__block uintmax_t i, n;
	__block tAutoLayoutParams pp;
	NSMutableString *constraint;
	BOOL firstStretchy;
	uintmax_t nStretchy;
	NSLayoutConstraintOrientation orientation;

	[self->v removeConstraints:[self->v constraints]];

	orientation = NSLayoutConstraintOrientationHorizontal;
	if (self->vertical)
		orientation = NSLayoutConstraintOrientationVertical;

	views = [NSMutableDictionary new];
	n = 0;
	[self->children enumerateObjectsUsingBlock:^(id obj, NSUInteger index, BOOL *stop) {
		id<tControl> c;
		NSNumber *isStretchy;
		NSLayoutPriority priority;

		c = (id<tControl>) obj;
		isStretchy = (NSNumber *) [self->stretchy objectAtIndex:n];
		// this also resets the hugging priority
		// TODO do this when adding and removing controls instead
		[c tFillAutoLayout:&pp];
		priority = NSLayoutPriorityDefaultHigh;			// forcibly hug; avoid stretching out
		if ([isStretchy boolValue])
			priority = NSLayoutPriorityDefaultLow;		// do not forcibly hug; freely stretch out
		[pp.view setContentHuggingPriority:priority forOrientation:orientation];
		[views setObject:pp.view forKey:tAutoLayoutKey(n)];
		n++;
	}];

	// first string the views together
	if (self->vertical)
		constraint = [NSMutableString stringWithString:@"V:|"];
	else
		constraint = [NSMutableString stringWithString:@"H:|"];
	firstStretchy = YES;
	for (i = 0; i < n; i++) {
		NSNumber *isStretchy;

		if (self->spaced && i != 0)
			[constraint appendString:@"-"];
		[constraint appendString:@"["];
		[constraint appendString:tAutoLayoutKey(i)];
		isStretchy = (NSNumber *) [self->stretchy objectAtIndex:i];
		if ([isStretchy boolValue])
			if (firstStretchy) {
				firstStretchy = NO;
				nStretchy = i;
			} else {
				[constraint appendString:@"(=="];
				[constraint appendString:tAutoLayoutKey(nStretchy)];
				[constraint appendString:@")"];
			}
		[constraint appendString:@"]"];
	}
	[constraint appendString:@"|"];
	[self->v addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:constraint options:0 metrics:nil views:views]];
	// TODO do not release constraint; it's autoreleased?

	// next make the views span the full other dimension
	for (i = 0; i < n; i++) {
		if (self->vertical)
			constraint = [NSMutableString stringWithString:@"H:|["];
		else
			constraint = [NSMutableString stringWithString:@"V:|["];
		[constraint appendString:tAutoLayoutKey(i)];
		[constraint appendString:@"]|"];
		[self->v addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:constraint options:0 metrics:nil views:views]];
		// TODO do not release constraint; it's autoreleased?
	}

	[views release];

	// and now populate for self
	p->view = self->v;
	p->attachLeft = YES;
	p->attachTop = YES;
	// don't attach to the end if there weren't any stretchy controls
	// firstStretchy is NO if there was at least one stretchy control
	if (self->vertical) {
		p->attachRight = YES;
		p->attachBottom = !firstStretchy;
	} else {
		p->attachRight = !firstStretchy;
		p->attachBottom = YES;
	}
}

- (void)tRelayout
{
	if (self->parent != nil)
		[self->parent tRelayout];
}

@end

File: redo/osxaltest/box.swift
// 7 august 2015
import Cocoa

struct BoxControl {
	var c: Control
	var stretchy: Bool
	var horzHuggingPri: NSLayoutPriority
	var vertHuggingPri: NSLayoutPriority
}

class Box : NSView, Control {
	private var controls: [BoxControl]
	private var parent: Control?
	private var vertical: Bool
	private var padded: Bool

	private var primaryDirPrefix: String
	private var secondaryDirPrefix: String
	private var primaryOrientation: NSLayoutConstraintOrientation
	private var secondaryOrientation: NSLayoutConstraintOrientation

	// we implement a lack of stretchy controls by adding a stretchy view at the end of the view list when we assemble layouts
	// this is that view
	private var noStretchyView: NSView

	init(vertical: Bool, padded: Bool) {
		self.controls = []
		self.parent = nil
		self.vertical = vertical
		self.padded = padded

		self.primaryDirPrefix = "H:"
		self.secondaryDirPrefix = "V:"
		self.primaryOrientation = NSLayoutConstraintOrientation.Horizontal
		self.secondaryOrientation = NSLayoutConstraintOrientation.Vertical
		if self.vertical {
			self.primaryDirPrefix = "V:"
			self.secondaryDirPrefix = "H:"
			self.primaryOrientation = NSLayoutConstraintOrientation.Vertical
			self.secondaryOrientation = NSLayoutConstraintOrientation.Horizontal
		}

		self.noStretchyView = NSView(frame: NSZeroRect)
		self.noStretchyView.translatesAutoresizingMaskIntoConstraints = false
		// make the view stretchy in both directions
		// you can tell this is correct by synthesizing an Add() in your head; see below
		setHorzHuggingPri(self.noStretchyView, myNSLayoutPriorityDefaultLow)
		setVertHuggingPri(self.noStretchyView, myNSLayoutPriorityDefaultLow)

		super.init(frame: NSZeroRect)
		self.translatesAutoresizingMaskIntoConstraints = false
	}

	required init?(coder: NSCoder) {
		fatalError("can't use this constructor, sorry")
	}

	func Add(control: Control, _ stretchy: Bool) {
		var c: BoxControl

		var view = control.View()
		c = BoxControl(
			c:				control,
			stretchy:			stretchy,
			horzHuggingPri:	horzHuggingPri(view),
			vertHuggingPri:	vertHuggingPri(view))
		self.addSubview(view)
		self.controls.append(c)

		// if a control is stretchy, it should not hug in the primary direction
		// otherwise, it should *forcibly* hug
		if c.stretchy {
			setHuggingPri(view, myNSLayoutPriorityDefaultLow, self.primaryOrientation)
		} else {
			// TODO will default high work?
			setHuggingPri(view, myNSLayoutPriorityRequired, self.primaryOrientation)
		}

		// make sure controls don't hug their secondary direction so they fill the width of the view
		setHuggingPri(view, myNSLayoutPriorityDefaultLow, self.secondaryOrientation)

		self.relayout()
	}

	func View() -> NSView {
		return self
	}

	func SetParent(p: Control) {
		self.parent = p
	}

	private func relayout() {
		var constraint: String

		if self.controls.count == 0 {
			return
		}

		self.removeConstraints(self.constraints)

		// first collect the views
		var views = [String: NSView]()
		var n = 0
		var firstStretchy = -1
		for c in self.controls {
			views["view\(n)"] = c.c.View()
			if firstStretchy == -1 && c.stretchy {
				firstStretchy = n
			}
			n++
		}

		// if there are no stretchy controls, we must add the no-stretchy view
		// if there are, we must remove it
		if firstStretchy == -1 {
			if self.noStretchyView.superview == nil {
				self.addSubview(self.noStretchyView)
			}
			views["noStretchyView"] = self.noStretchyView
		} else {
			if self.noStretchyView.superview != nil {
				self.noStretchyView.removeFromSuperview()
			}
		}

		// next, assemble the views in the primary direction
		// they all go in a straight line
		constraint = "\(self.primaryDirPrefix)|"
		for i in 0..<n {
			if self.padded && i != 0 {
				constraint += "-"
			}
			constraint += "[view\(i)"
			// implement multiple stretchiness properly
			if self.controls[i].stretchy && i != firstStretchy {
				constraint += "(==view\(firstStretchy))"
			}
			constraint += "]"
		}
		if firstStretchy == -1 {		// don't space between the last control and the no-stretchy view
			constraint += "[noStretchyView]"
		}
		constraint += "|"
		var constraints = mkconstraints(constraint, nil, views)
		self.addConstraints(constraints)

		// next: assemble the views in the secondary direction
		// each of them will span the secondary direction
		for i in 0..<n {
			constraint = "\(self.secondaryDirPrefix)|[view\(i)]|"
			var constraints = mkconstraints(constraint, nil, views)
			self.addConstraints(constraints)
		}
		if firstStretchy == -1 {		// and again to the no-stretchy view
			constraint = "\(self.secondaryDirPrefix)|[noStretchyView]|"
			var constraints = mkconstraints(constraint, nil, views)
			self.addConstraints(constraints)
		}
	}
}
